src/components/tooltip/GCMarker.js

```diff
@@ -423,80 +430,104 @@ function _treeInsert(
-    if (node.value) {
-      // We delete the value to change this node from a leaf to a branch.
-      delete node.value;
-    }
+    node.leaf = false;
     _treeInsert(node.branches, path, phase);
   } else {
     // Make the new node leaf node.
-    if (node.value) {
+    if (node.value.name !== '(temp)') {
       console.error(
-        'Duplicate phases in _treeInsert in MarkerTooltipContents.js'
+        `Duplicate phase ${phase.name} in _treeInsert in MarkerTooltipContents.js`
       );
       return;
     }
     node.value = phase;
+    node.leaf = true;
   }
 }
 
-function _treeGetLeaves(
-  tree: Map<string, PhaseTreeNode>
-): Array<PhaseTimeTuple> {
-  const leaves = [];
+function _treeMapNodes<T>(
+  tree: Map<string, PhaseTreeNode>,
+  f: (node: PhaseTreeNode) => T | null,
+  results: Array<T> = []
+): Array<T> {
   for (const node of tree.values()) {
-    if (node.value) {
-      leaves.push(node.value);
-    } else {
-      leaves.push(..._treeGetLeaves(node.branches));
+    const result = f(node);
+    if (result !== null) {
+      results.push(result);
+    }
+    if (!node.leaf) {
+      _treeMapNodes(node.branches, f, results);
     }
   }
-  return leaves;
+  return results;
 }
 
 function _filterInterestingPhaseTimes(
   rawPhases: PhaseTimes<Microseconds>,
   numSelect: number
 ): Array<PhaseTimeTuple> {
-  let phaseTimes = _makePhaseTimesArray(rawPhases);
+  const phaseTimes = _makePhaseTimesArray(rawPhases);
 
   /*
-   * Select only the leaf phases.
+   * Build the tree.
    */
   const tree = new Map();
   for (const phase of phaseTimes) {
     const components = phase.name.split('.');
     _treeInsert(tree, components, phase);
   }
-  phaseTimes = _treeGetLeaves(tree);
 
   /*
-   * Of those N leaf phases, select the M most interesting phases by
-   * determining the threshold we want to stop including phases at.
-   *
-   * Calculate the threshold by sorting the list of times in asscending
-   * order, then slicing off all the low items and looking at the first
-   * item.
+   * For every non-leaf phase, add in a (self-time) leaf with the time not
+   * accounted for by its children.
    */
-  const sortedPhaseTimes = phaseTimes
-    .map(pt => pt.time)
-    // Descending order
-    .sort((a, b) => b - a);
-  const threshold = sortedPhaseTimes[numSelect];
+  _treeMapNodes(
+    tree,
+    ({ leaf, branches, value: { time: totalTime, name } }) => {
+      if (!leaf) {
+        let remaining = totalTime;
+        for (const { value } of branches.values()) {
+          remaining -= value.time;
+        }
+        const selfTimeName = `${name} (self-time)`;
+        branches.set(name, {
+          branches: new Map(),
+          leaf: true,
+          value: { name: selfTimeName, time: remaining },
+        });
+      }
+      return null;
+    }
+  );
+
+  // Select the N most interesting leaf phases, in order of descending time.
+  const sortedPhaseTimes = _treeMapNodes(tree, node =>
+    node.leaf ? node.value : null
```

Comment:
I'm not sure why this ternary, especially why we use `node.value` here?

BTW if you follow my suggestion above, then the filter function could return a "boolean-ish" value. Flow does [`typeof Boolean`](https://github.com/facebook/flow/blob/a9c0df5b6459f3635aab7ffa40f2b21e997e74a3/lib/core.js#L707) but you can specify that it returns a true boolean if you prefer.


Note: the function parameter for the "loop" function would return `mixed` (means we don't care about its return value).

Reply:
These are all kind of the same thing. I guess the way I had it feels a little more direct -- I'm not grabbing out a list of tree nodes (`PhaseTreeNode`), I'm getting a list of phase times (`PhaseTimeTuple`), one for each leaf node. The "for each leaf node" part is the reason for the ternary; I explicitly only want leaves for the reasons explained in Paul's comment. But I don't want the leaves themselves, I want their payloads, and that's the reason for returning `node.value` instead of using a boolean.

I think of `map` as something that you can do to a general data structure, not just an array/list. "Visit all nodes and return the results of applying a function, in order of visitation." But mathematically, you `map` one sequence to another, so perhaps I'm just being weird. Having written it out, now I think that `map` on a tree should return a tree somehow.

I can do this in multiple steps: first select all the leaves, then map them to their values, and only then sort and extract a sublist. @PaulBone's original `_treeGetLeaves` did the first two steps, so I'll just switch back to that. Then I won't need the general utility `_flattenTreeWithFiltering` at all, I just need to add `_forEachTreeNode`.